pubsub/nats: data race found by Travis

```
==================
WARNING: DATA RACE
Read at 0x00c0001336e0 by goroutine 39:
  github.com/nats-io/go-nats.(*Conn).close()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/nats.go:3482 +0x380
  github.com/nats-io/go-nats.(*Conn).Close()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/nats.go:3518 +0x46
  github.com/nats-io/go-nats.(*Conn).processErr()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/nats.go:2388 +0x210
  github.com/nats-io/go-nats.(*Conn).parse()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/parser.go:255 +0xbb7
  github.com/nats-io/go-nats.(*Conn).readLoop()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/nats.go:1983 +0x248
Previous write at 0x00c0001336e0 by goroutine 78:
  github.com/nats-io/go-nats.(*Conn).SubscribeSync()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/nats.go:2744 +0x128
  gocloud.dev/pubsub/natspubsub.createSubscription()
      /home/travis/gopath/src/gocloud.dev/pubsub/natspubsub/nats.go:175 +0x68
  gocloud.dev/pubsub/natspubsub.CreateSubscription()
      /home/travis/gopath/src/gocloud.dev/pubsub/natspubsub/nats.go:171 +0x68
  gocloud.dev/pubsub/natspubsub.TestBadSubjects()
      /home/travis/gopath/src/gocloud.dev/pubsub/natspubsub/nats_test.go:291 +0x176
  testing.tRunner()
      /home/travis/.gimme/versions/go1.11.5.linux.amd64/src/testing/testing.go:827 +0x162
Goroutine 39 (running) created at:
  github.com/nats-io/go-nats.(*Conn).processConnectInit()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/nats.go:1369 +0x299
  github.com/nats-io/go-nats.(*Conn).connect()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/nats.go:1393 +0x3b4
  github.com/nats-io/go-nats.Options.Connect()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/nats.go:955 +0x3a9
  github.com/nats-io/go-nats.Connect()
      /home/travis/gopath/pkg/mod/github.com/nats-io/go-nats@v1.7.2/nats.go:521 +0x2b0
  gocloud.dev/pubsub/natspubsub.newHarness()
      /home/travis/gopath/src/gocloud.dev/pubsub/natspubsub/nats_test.go:48 +0x168
  gocloud.dev/pubsub/natspubsub.TestBadSubjects()
      /home/travis/gopath/src/gocloud.dev/pubsub/natspubsub/nats_test.go:284 +0x7d
  testing.tRunner()
      /home/travis/.gimme/versions/go1.11.5.linux.amd64/src/testing/testing.go:827 +0x162
Goroutine 78 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.11.5.linux.amd64/src/testing/testing.go:878 +0x659
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.11.5.linux.amd64/src/testing/testing.go:1119 +0xa8
  testing.tRunner()
      /home/travis/.gimme/versions/go1.11.5.linux.amd64/src/testing/testing.go:827 +0x162
  testing.runTests()
      /home/travis/.gimme/versions/go1.11.5.linux.amd64/src/testing/testing.go:1117 +0x4ee
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.11.5.linux.amd64/src/testing/testing.go:1034 +0x2ee
  main.main()
      _testmain.go:644 +0x342
==================
--- FAIL: TestBadSubjects (0.03s)
    testing.go:771: race detected during execution of test
FAIL
```

--- a/internal/testing/alldeps
+++ b/internal/testing/alldeps
@@ -61,6 +61,7 @@ github.com/mitchellh/go-homedir
 github.com/mitchellh/go-testing-interface
 github.com/mitchellh/mapstructure
 github.com/nats-io/go-nats
+github.com/nats-io/nats.go
 github.com/nats-io/nkeys
 github.com/nats-io/nuid
 github.com/oklog/run

--- a/internal/website/data/examples.json
+++ b/internal/website/data/examples.json
@@ -160,11 +160,11 @@
 		"code": "topic, err := pubsub.OpenTopic(ctx, \"mem://topicA\")\nif err != nil {\n\treturn err\n}\ndefer topic.Shutdown(ctx)"
 	},
 	"gocloud.dev/pubsub/natspubsub.ExampleOpenSubscription": {
-		"imports": "import (\n\t\"context\"\n\n\t\"github.com/nats-io/go-nats\"\n\t\"gocloud.dev/pubsub/natspubsub\"\n)",
+		"imports": "import (\n\t\"context\"\n\n\t\"github.com/nats-io/nats.go\"\n\t\"gocloud.dev/pubsub/natspubsub\"\n)",
 		"code": "natsConn, err := nats.Connect(\"nats://nats.example.com\")\nif err != nil {\n\treturn err\n}\ndefer natsConn.Close()\n\nsubscription, err := natspubsub.OpenSubscription(\n\tnatsConn,\n\t\"example.mysubject\",\n\tnil)\nif err != nil {\n\treturn err\n}\ndefer subscription.Shutdown(ctx)"
 	},
 	"gocloud.dev/pubsub/natspubsub.ExampleOpenTopic": {
-		"imports": "import (\n\t\"context\"\n\n\t\"github.com/nats-io/go-nats\"\n\t\"gocloud.dev/pubsub/natspubsub\"\n)",
+		"imports": "import (\n\t\"context\"\n\n\t\"github.com/nats-io/nats.go\"\n\t\"gocloud.dev/pubsub/natspubsub\"\n)",
 		"code": "natsConn, err := nats.Connect(\"nats://nats.example.com\")\nif err != nil {\n\treturn err\n}\ndefer natsConn.Close()\n\ntopic, err := natspubsub.OpenTopic(natsConn, \"example.mysubject\", nil)\nif err != nil {\n\treturn err\n}\ndefer topic.Shutdown(ctx)"
 	},
 	"gocloud.dev/pubsub/natspubsub.Example_openSubscriptionFromURL": {

--- a/pubsub/natspubsub/example_test.go
+++ b/pubsub/natspubsub/example_test.go
@@ -18,7 +18,7 @@ import (
 	"context"
 	"log"
 
-	"github.com/nats-io/go-nats"
+	"github.com/nats-io/nats.go"
 	"gocloud.dev/pubsub"
 	"gocloud.dev/pubsub/natspubsub"
 )

--- a/pubsub/natspubsub/go.mod
+++ b/pubsub/natspubsub/go.mod
@@ -16,9 +16,7 @@ module gocloud.dev/pubsub/natspubsub
 
 require (
 	github.com/google/go-cmp v0.3.0
-	github.com/nats-io/gnatsd v1.4.1
-	github.com/nats-io/go-nats v1.7.2
-	github.com/nats-io/nkeys v0.0.2 // indirect
-	github.com/nats-io/nuid v1.0.1 // indirect
+	github.com/nats-io/nats-server/v2 v2.0.0
+	github.com/nats-io/nats.go v1.8.1
 	gocloud.dev v0.15.0
 )

--- a/pubsub/natspubsub/nats.go
+++ b/pubsub/natspubsub/nats.go
@@ -55,7 +55,7 @@ import (
 	"sync"
 	"time"
 
-	"github.com/nats-io/go-nats"
+	"github.com/nats-io/nats.go"
 	"gocloud.dev/gcerrors"
 	"gocloud.dev/internal/batcher"
 	"gocloud.dev/pubsub"

--- a/pubsub/natspubsub/nats_test.go
+++ b/pubsub/natspubsub/nats_test.go
@@ -27,9 +27,9 @@ import (
 	"gocloud.dev/pubsub/drivertest"
 
 	"github.com/google/go-cmp/cmp"
-	"github.com/nats-io/gnatsd/server"
-	gnatsd "github.com/nats-io/gnatsd/test"
-	"github.com/nats-io/go-nats"
+	"github.com/nats-io/nats-server/v2/server"
+	gnatsd "github.com/nats-io/nats-server/v2/test"
+	"github.com/nats-io/nats.go"
 )
 
 const (
@@ -202,6 +202,10 @@ func TestInteropWithDirectNATS(t *testing.T) {
 		t.Fatal(err)
 	}
 	msg, err := ps.Receive(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer msg.Ack()
 	if !bytes.Equal(msg.Body, body) {
 		t.Fatalf("Data did not match. %q vs %q\n", m.Data, body)
 	}
@@ -275,8 +279,6 @@ func TestErrorCode(t *testing.T) {
 	}
 }
 
-/* Temporarily disabled due to #1556, a data race in NATS.
-
 func TestBadSubjects(t *testing.T) {
 	ctx := context.Background()
 	dh, err := newHarness(ctx, t)
@@ -286,7 +288,7 @@ func TestBadSubjects(t *testing.T) {
 	defer dh.Close()
 	h := dh.(*harness)
 
-	sub, err := CreateSubscription(h.nc, "..bad", func() { t.Fatal("ack called unexpectedly") }, nil)
+	sub, err := OpenSubscription(h.nc, "..bad", nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -294,7 +296,7 @@ func TestBadSubjects(t *testing.T) {
 		t.Fatal("Expected an error with bad subject")
 	}
 
-	pt, err := CreateTopic(h.nc, "..bad", nil)
+	pt, err := OpenTopic(h.nc, "..bad", nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -302,7 +304,6 @@ func TestBadSubjects(t *testing.T) {
 		t.Fatal("Expected an error with bad subject")
 	}
 }
-*/
 
 func BenchmarkNatsPubSub(b *testing.B) {
 	ctx := context.Background()

